src/lib/storage/table.cpp

```diff
@@ -95,6 +94,36 @@ Chunk& Table::get_chunk(ChunkID chunk_id) {
 
 const Chunk& Table::get_chunk(ChunkID chunk_id) const { return get_chunk(chunk_id); }
 
-void Table::compress_chunk(ChunkID chunk_id) { throw std::runtime_error("Implement Table::compress_chunk"); }
+void Table::compress_chunk(ChunkID chunk_id) {
+  const auto& old_chunk = get_chunk(chunk_id);
+
+  // new empty chunk
+  auto new_chunk = std::make_unique<Chunk>();
+
+  std::vector<std::thread> threads;
+  threads.reserve(old_chunk.size());
+
+  for (ColumnID column_id(0); column_id < old_chunk.size(); ++column_id) {
+    const auto base_segment = old_chunk.get_segment(column_id);
+    const auto segment_type = _column_types[column_id];
+
+    std::shared_ptr<BaseSegment> compressed_segment;
+
+    threads.push_back(std::thread([&compressed_segment, &base_segment, &segment_type]() {
+      // build dictionary compressed segment
+      compressed_segment = make_shared_by_data_type<BaseSegment, DictionarySegment>(segment_type, base_segment);
+    }));
+
+    // add segment to chunk
+    new_chunk->add_segment(compressed_segment);
```

Comment:
It took me a while to understand why this is correct - I think the current comments in this method are not very useful, an explanation why setting the `shared_ptr` inside the thread (that probably executes after the shared_ptr has been added to the chunk) is correct would be much more useful.

Reply:
valid criticism, the comments are really bad and we are not sure if these threads really do what they should - we'll replace this with promises

Reply:
>  took me a while to understand why this is correct

Help me. How is it guaranteed that `compressed_segment` is not `nullptr` at this point?